File: apps/api/src/domain/date.ts
export function parseDateOnly(value: string | Date): Date {
  if (value instanceof Date) {
    return new Date(Date.UTC(value.getUTCFullYear(), value.getUTCMonth(), value.getUTCDate()));
  }

  const trimmed = value.trim();
  const dateOnlyMatch = /^\d{4}-\d{2}-\d{2}$/.test(trimmed);
  const isoDateTimeMatch = /^\d{4}-\d{2}-\d{2}[T\s]/.test(trimmed);

  if (!dateOnlyMatch && !isoDateTimeMatch) {
    throw new Error('Date must use YYYY-MM-DD format');
  }

  const dateKey = trimmed.slice(0, 10);
  return new Date(`${dateKey}T00:00:00.000Z`);
}

export function toDateKey(value: string | Date): string {
  const date = parseDateOnly(value);
  return date.toISOString().slice(0, 10);
}

export function addDays(value: Date, days: number): Date {
  const date = parseDateOnly(value);
  date.setUTCDate(date.getUTCDate() + days);
  return date;
}

export function eachDay(startDate: Date, endDate: Date): Date[] {
  const start = parseDateOnly(startDate);
  const end = parseDateOnly(endDate);
  const days: Date[] = [];

  for (let cursor = start; cursor <= end; cursor = addDays(cursor, 1)) {
    days.push(cursor);
  }

  return days;
}

export function rangesOverlap(
  firstStart: Date,
  firstEnd: Date,
  secondStart: Date,
  secondEnd: Date
) {
  return firstStart <= secondEnd && firstEnd >= secondStart;
}

export function isWeekend(value: Date) {
  const day = value.getUTCDay();
  return day === 0 || day === 6;
}

export function workingDaysInRange(startDate: Date, endDate: Date) {
  return eachDay(startDate, endDate).filter((day) => !isWeekend(day));
}

export function monthRange(month: string) {
  if (!/^\d{4}-\d{2}$/.test(month)) {
    throw new Error('Month must use YYYY-MM format');
  }

  const [year, monthIndex] = month.split('-').map(Number);
  const startDate = new Date(Date.UTC(year, monthIndex - 1, 1));
  const endDate = new Date(Date.UTC(year, monthIndex, 0));

  return { startDate, endDate };
}

File: apps/api/cmd/api/date.go
package main

import (
	"fmt"
	"time"
)

func parseDateOnly(value string) (time.Time, error) {
	if len(value) < 10 {
		return time.Time{}, fmt.Errorf("date must use YYYY-MM-DD format")
	}
	key := value[:10]
	parsed, err := time.ParseInLocation("2006-01-02", key, time.UTC)
	if err != nil {
		return time.Time{}, fmt.Errorf("date must use YYYY-MM-DD format")
	}
	return parsed, nil
}

func toDateKey(value time.Time) string {
	return value.UTC().Format("2006-01-02")
}

func addDays(value time.Time, days int) time.Time {
	return value.UTC().AddDate(0, 0, days)
}

func eachDay(startDate, endDate time.Time) []time.Time {
	start := normalizeDate(startDate)
	end := normalizeDate(endDate)
	days := make([]time.Time, 0)
	for cursor := start; !cursor.After(end); cursor = addDays(cursor, 1) {
		days = append(days, cursor)
	}
	return days
}

func isWeekend(value time.Time) bool {
	switch value.UTC().Weekday() {
	case time.Saturday, time.Sunday:
		return true
	default:
		return false
	}
}

func workingDaysInRange(startDate, endDate time.Time) []time.Time {
	all := eachDay(startDate, endDate)
	out := make([]time.Time, 0, len(all))
	for _, day := range all {
		if !isWeekend(day) {
			out = append(out, day)
		}
	}
	return out
}

func monthRange(month string) (time.Time, time.Time, error) {
	if len(month) != 7 {
		return time.Time{}, time.Time{}, fmt.Errorf("month must use YYYY-MM format")
	}
	parsed, err := time.ParseInLocation("2006-01", month, time.UTC)
	if err != nil {
		return time.Time{}, time.Time{}, fmt.Errorf("month must use YYYY-MM format")
	}
	start := time.Date(parsed.Year(), parsed.Month(), 1, 0, 0, 0, 0, time.UTC)
	end := start.AddDate(0, 1, -1)
	return start, end, nil
}

func normalizeDate(value time.Time) time.Time {
	return time.Date(value.UTC().Year(), value.UTC().Month(), value.UTC().Day(), 0, 0, 0, 0, time.UTC)
}
